File: sdk/typescript/cortexdb-client/grounding.ts
import type { AnswerGroundingReportResponse, ContextPackResponse, GroundedAnswerResponse, VerificationReportResponse } from "./types";

function uniqueValues<T>(values: T[]): T[] {
  const seen = new Set<T>();
  const out: T[] = [];
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      out.push(value);
    }
  }
  return out;
}

function tokenize(text: string): string[] {
  const stopwords = new Set(["a", "an", "and", "the", "or", "of", "to", "in"]);
  const terms = text
    .toLowerCase()
    .split(/[^\p{L}\p{N}]+/u)
    .filter((term) => term.length > 0 && !stopwords.has(term));
  return [...new Set(terms)].sort();
}

function splitAnswerSpans(answer: string): Array<[string, number, number]> {
  const spans: Array<[string, number, number]> = [];
  let start = 0;
  for (let index = 0; index < answer.length; index += 1) {
    const ch = answer[index];
    const decimalDot = ch === "." && /\d/.test(answer[index - 1] ?? "") && /\d/.test(answer[index + 1] ?? "");
    if (ch === "!" || ch === "?" || ch === "\n" || (ch === "." && !decimalDot)) {
      pushAnswerSpan(answer, start, index + 1, spans);
      start = index + 1;
    }
  }
  pushAnswerSpan(answer, start, answer.length, spans);
  return spans;
}

function pushAnswerSpan(
  answer: string,
  start: number,
  end: number,
  spans: Array<[string, number, number]>,
): void {
  const raw = answer.slice(start, end);
  const text = raw.trim();
  if (text.length === 0) return;
  const leading = raw.length - raw.trimStart().length;
  const trailing = raw.length - raw.trimEnd().length;
  spans.push([text, start + leading, end - trailing]);
}

function q16Ratio(numerator: number, denominator: number): number {
  if (denominator === 0) return 65535;
  return Math.floor((numerator * 65535) / denominator);
}

export function groundAnswer(
  context: ContextPackResponse,
  answer: string,
  options: {
    minSpanSupportQ16?: number;
    requireCitations?: boolean;
    rejectUnsupported?: boolean;
  } = {},
): AnswerGroundingReportResponse {
  const minSpanSupportQ16 = options.minSpanSupportQ16 ?? 65535;
  const requireCitations = options.requireCitations ?? false;
  const rejectUnsupported = options.rejectUnsupported ?? false;
  const spans = splitAnswerSpans(answer).map(([text, start, end]) => {
    const spanTerms = tokenize(text);
    if (spanTerms.length === 0) {
      return {
        text,
        start_byte: start,
        end_byte: end,
        support_q16: 65535,
        supported: true,
        covered_terms: [],
        missing_terms: [],
        supported_by_cell_ids: [],
        citations: [],
      };
    }
    const covered = new Set<string>();
    const cellIds: number[] = [];
    const citations: string[] = [];
    for (const cell of context.cells) {
      const cellTerms = new Set(tokenize(cell.payload_text));
      let matched = false;
      for (const term of spanTerms) {
        if (cellTerms.has(term)) {
          covered.add(term);
          matched = true;
        }
      }
      if (matched) {
        cellIds.push(cell.cell_id);
        if (cell.citation) citations.push(cell.citation);
      }
    }
    const support = q16Ratio(covered.size, spanTerms.length);
    const supported = support >= minSpanSupportQ16 && (!requireCitations || citations.length > 0);
    return {
      text,
      start_byte: start,
      end_byte: end,
      support_q16: support,
      supported,
      covered_terms: [...covered].sort(),
      missing_terms: spanTerms.filter((term) => !covered.has(term)),
      supported_by_cell_ids: uniqueValues(cellIds),
      citations: uniqueValues(citations),
    };
  });
  const supportedSpanCount = spans.filter((span) => span.supported).length;
  const unsupportedSpanCount = spans.length - supportedSpanCount;
  const support = spans.length === 0
    ? 65535
    : Math.floor(spans.reduce((total, span) => total + span.support_q16, 0) / spans.length);
  return {
    answer_supported: unsupportedSpanCount === 0,
    rejected: rejectUnsupported && unsupportedSpanCount > 0,
    support_q16: support,
    supported_span_count: supportedSpanCount,
    unsupported_span_count: unsupportedSpanCount,
    spans,
  };
}

export function buildGroundedAnswerResponse(params: {
  question: string;
  answer: string;
  retrieveStatement: string;
  verifyStatement?: string | null;
  context: ContextPackResponse;
  verification?: VerificationReportResponse | null;
  requireCitations?: boolean;
  minSpanSupportQ16?: number;
  rejectUnsupported?: boolean;
}): GroundedAnswerResponse {
  const grounding = groundAnswer(params.context, params.answer, {
    requireCitations: params.requireCitations,
    minSpanSupportQ16: params.minSpanSupportQ16,
    rejectUnsupported: params.rejectUnsupported,
  });
  const citations = uniqueValues([
    ...grounding.spans.flatMap((span) => span.citations),
    ...params.context.cells.flatMap((cell) => cell.citation ? [cell.citation] : []),
  ]);
  const usedContextCellIds = uniqueValues([
    ...grounding.spans.flatMap((span) => span.supported_by_cell_ids),
    ...params.context.cells.map((cell) => cell.cell_id),
  ]);
  return {
    question: params.question,
    answer: params.answer,
    retrieve_statement: params.retrieveStatement,
    verify_statement: params.verifyStatement ?? null,
    context: params.context,
    grounding,
    verification: params.verification ?? null,
    citations,
    used_context_cell_ids: usedContextCellIds,
    rejected: grounding.rejected,
  };
}

File: sdk/python/_cortexdb_client/grounding.py
from __future__ import annotations

from typing import Any

from .models import (
    AnswerGroundingReportResponse,
    AnswerGroundingSpanResponse,
    ContextPackResponse,
    GroundedAnswerResponse,
    VerificationReportResponse,
)


def _tokenize(text: str) -> tuple[str, ...]:
    terms: list[str] = []
    current: list[str] = []
    for character in text.lower():
        if character.isalnum():
            current.append(character)
        elif current:
            term = "".join(current)
            if term not in {"a", "an", "and", "the", "or", "of", "to", "in"}:
                terms.append(term)
            current = []
    if current:
        term = "".join(current)
        if term not in {"a", "an", "and", "the", "or", "of", "to", "in"}:
            terms.append(term)
    return tuple(sorted(set(terms)))


def _split_answer_spans(answer: str) -> tuple[tuple[str, int, int], ...]:
    spans: list[tuple[str, int, int]] = []
    start = 0
    for index, character in enumerate(answer):
        if character in {"!", "?", "\n"} or (
            character == "."
            and not (
                index > 0
                and index + 1 < len(answer)
                and answer[index - 1].isdigit()
                and answer[index + 1].isdigit()
            )
        ):
            _push_answer_span(answer, start, index + 1, spans)
            start = index + 1
    _push_answer_span(answer, start, len(answer), spans)
    return tuple(spans)


def _push_answer_span(
    answer: str,
    start: int,
    end: int,
    spans: list[tuple[str, int, int]],
) -> None:
    raw = answer[start:end]
    text = raw.strip()
    if not text:
        return
    leading = len(raw) - len(raw.lstrip())
    trailing = len(raw) - len(raw.rstrip())
    spans.append((text, start + leading, end - trailing))


def _q16_ratio(numerator: int, denominator: int) -> int:
    if denominator == 0:
        return 65535
    return int(numerator * 65535 / denominator)


def _unique(values: list[str] | list[int]) -> tuple[Any, ...]:
    seen = set()
    out = []
    for value in values:
        if value not in seen:
            seen.add(value)
            out.append(value)
    return tuple(out)


def ground_answer(
    context: ContextPackResponse,
    answer: str,
    *,
    min_span_support_q16: int,
    require_citations: bool,
    reject_unsupported: bool,
) -> AnswerGroundingReportResponse:
    spans: list[AnswerGroundingSpanResponse] = []
    for text, start, end in _split_answer_spans(answer):
        span_terms = _tokenize(text)
        if not span_terms:
            spans.append(
                AnswerGroundingSpanResponse(
                    text=text,
                    start_byte=start,
                    end_byte=end,
                    support_q16=65535,
                    supported=True,
                    covered_terms=(),
                    missing_terms=(),
                    supported_by_cell_ids=(),
                    citations=(),
                )
            )
            continue
        covered: set[str] = set()
        cell_ids: list[int] = []
        citations: list[str] = []
        for cell in context.cells:
            cell_terms = set(_tokenize(cell.payload_text))
            matched = False
            for term in span_terms:
                if term in cell_terms:
                    covered.add(term)
                    matched = True
            if matched:
                cell_ids.append(cell.cell_id)
                if cell.citation:
                    citations.append(cell.citation)
        support = _q16_ratio(len(covered), len(span_terms))
        supported = support >= min_span_support_q16 and (
            not require_citations or bool(citations)
        )
        spans.append(
            AnswerGroundingSpanResponse(
                text=text,
                start_byte=start,
                end_byte=end,
                support_q16=support,
                supported=supported,
                covered_terms=tuple(sorted(covered)),
                missing_terms=tuple(term for term in span_terms if term not in covered),
                supported_by_cell_ids=_unique(cell_ids),
                citations=_unique(citations),
            )
        )
    supported_count = sum(1 for span in spans if span.supported)
    unsupported_count = len(spans) - supported_count
    average = int(sum(span.support_q16 for span in spans) / len(spans)) if spans else 65535
    return AnswerGroundingReportResponse(
        answer_supported=unsupported_count == 0,
        rejected=reject_unsupported and unsupported_count > 0,
        support_q16=average,
        supported_span_count=supported_count,
        unsupported_span_count=unsupported_count,
        spans=tuple(spans),
    )


def _grounded_answer_response(
    *,
    question: str,
    answer: str,
    retrieve_statement: str,
    verify_statement: str | None,
    context: ContextPackResponse,
    verification: "VerificationReportResponse | None",
    require_citations: bool,
    reject_unsupported: bool,
) -> GroundedAnswerResponse:
    grounding = context.ground_answer(
        answer,
        require_citations=require_citations,
        reject_unsupported=reject_unsupported,
    )
    citations = _unique(
        [
            citation
            for span in grounding.spans
            for citation in span.citations
        ]
        + [cell.citation for cell in context.cells if cell.citation]
    )
    used_cell_ids = _unique(
        [
            cell_id
            for span in grounding.spans
            for cell_id in span.supported_by_cell_ids
        ]
        + [cell.cell_id for cell in context.cells]
    )
    return GroundedAnswerResponse(
        question=question,
        answer=answer,
        retrieve_statement=retrieve_statement,
        verify_statement=verify_statement,
        context=context,
        grounding=grounding,
        verification=verification,
        citations=citations,
        used_context_cell_ids=used_cell_ids,
        rejected=grounding.rejected,
    )


